### PR TITLE
Fix Gradle warning in a freshly `flutter create`ed Android project

### LINKED ISSUE #122289: Top-level `build.gradle` – Consider using 'tasks.register' to avoid unnecessary configuration
## Steps to Reproduce

1. `flutter create` a new project with Android support.
2. Open the Android project in Android Studio/IntellIJ
3. Go to top-level `build.gradle` file and see warning with the `clean` Gradle task declaration:

    Screenshot 2023-03-09 at 1 31 42 PM

    Screenshot 2023-03-09 at 1 31 48 PM

**Expected results:**

No warnings.

**Actual results:**

The warning is present, which doesn't look good since the project has just been generated using the official tool.

**Environment**

$ flutter --version
Flutter 3.7.7 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 2ad6cd72c0 (19 hours ago) • 2023-03-08 09:41:59 -0800
Engine • revision 1837b5be5f
Tools • Dart 2.19.4 • DevTools 2.20.1

## Fix 

It's pretty simple. Replace:

```groovy
task clean(type: Delete) {
    delete rootProject.buildDir
}
```

with:

```
tasks.register("clean", Delete) {
    delete rootProject.buildDir
}
```

### PR DESCRIPTION
**Pull request by bartekpacia**
This PR fixes #122289

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by gnprice** [comment]
Neat!

Can you try adding a migration for this?
https://github.com/flutter/flutter/tree/main/packages/flutter_tools/lib/src/migrations
https://github.com/flutter/flutter/tree/main/packages/flutter_tools/lib/src/ios/migrations
https://github.com/flutter/flutter/wiki/Flutter-migrate

That way (a) existing Flutter projects will get the benefit too, (b) the divergence between Flutter projects created at different times can be reduced.

**Comment by bartekpacia** [comment]
Thanks for taking a look and for these links :)

Even though I consider myself a fairly experienced Flutter developer, I have never used `flutter migrate` on the CLI 😅 Maybe an IDE did some of it for me, but I can't recall. I feel like `flutter migrate` is not marketed enough.

The migration sounds like a good idea, but I don't think it's worth it for this change. After all, it's a very minor change and I'm pretty sure the performance benefit is negligible. People who'll get the warning will Google it and replace.

That said, if you think it's worth adding this anyway, I'll be happy to dive in and write the migration when I find some free time :)

**Comment by gnprice** [comment]
> Even though I consider myself a fairly experienced Flutter developer, I have never used `flutter migrate` on the CLI  Maybe an IDE did some of it for me, but I can't recall. I feel like `flutter migrate` is not advertised enough - which is a pity, because it looks nice.

Yeah, this appears to be an entirely separate system from `flutter migrate`. (Which itself apparently no longer exists? Try `flutter migrate --help` and it says it can't find the command.)

If you take one of those existing migration classes and look at where it's used, you find that the migrations are applied as basically the first step of `flutter run` or `flutter build`. For example, the `buildLinux` function has this near the top:
```dart
  final List<ProjectMigrator> migrators = <ProjectMigrator>[
    CmakeCustomCommandMigration(linuxProject, globals.logger),
  ];

  final ProjectMigration migration = ProjectMigration(migrators);
  migration.run();
```
before it goes on to call `_runCmake` and `_runBuild`, and that `buildLinux` function is in turn called from `flutter build linux` and `flutter run` for a Linux target.

So this system doesn't need any advertising; it's already integrated into people's normal workflows.



> The migration sounds like a good idea, but I don't think it's worth it for this change. After all, it's a very minor change and I'm pretty sure the performance benefit is negligible. People who'll get the warning will Google it and replace the default, eager `clean` task declaration with a lazy one.

Everyone will get the warning when they upgrade to Gradle 8+, right? Everyone who's building for Android, anyway.

Many Flutter developers will make that upgrade themselves after Gradle 8 reaches a stable release. And then in a few years everyone will make that upgrade who hasn't already, with the successor to #122376 when some Gradle 8+ version is needed for Java 21.

That's a lot of people who can be saved the effort of spotting this warning, figuring out what it's talking about, locating the fix, and then applying it and convincing themselves it was correct and didn't break anything. (Especially as many of those developers will know very little about Gradle — I suspect that even most people writing native Android apps try to avoid thinking about Gradle much, and all the more so for people working mainly in Flutter.)

So if you're up for taking the understanding you've already acquired of what this warning is about, and spending a bit of time to write and test an automated migration that encodes that understanding, then I think that would be a quite useful thing to do.

It looks like this would also be the pioneering example of a migration for Android builds — all the other platforms have one or more (and iOS is in the lead with ten), but not Android. Fortunately I think that part will be straightforward, just adding a few lines at the top of `buildGradleApp` similar to the code I quoted above.

**Comment by bartekpacia** [comment]
Thank you very much for this thorough explanation @gnprice. I agree with this - let's spare people some Gradle warnings :)

I'll add the migration code to this PR when I find some free time (in a few days).

**Comment by gnprice** [comment]
Great!

If you find you have any questions about how that migration system works, I recommend you go to the tools channel in Discord and ping jmagman there. She's the expert on this system, and I've seen her several times recently recommend people use it for various changes.

**Comment by bartekpacia** [comment]
Hi @gnprice, I added a migrator :) I'll be happy if you review!

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:95]: Another `FakeAndroidProject` class already exists [here](https://github.com/flutter/flutter/blob/cd19817d3fdbd89646710a4f09fcd9b2acfe8a0b/packages/flutter_tools/test/general.shard/plugins_test.dart#L1805).

I think it's okay to duplicate it because it looks like they exist for different purposes, but I'd rather ask you :)

Comment (gnprice) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:95]: I agree. It's not even really duplicating so much as it is two different tests using the same functionality, and making up the same reasonable name for their local private class that uses it.

(It's not private through the Dart underscore-prefix mechanism, but it's nevertheless effectively private because it's defined in a file that defines `main` and won't be imported.)

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:97]: This will be the root (aka `/`) of the simulated filesystem. Probably cleaner to put it a bit deeper, like an `android` subdirectory. Potentially in the future something else in this group of tests will want to put something else in the test filesystem.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:23]: This doesn't seem to really test anything, except to some extent the general `ProjectMigration` class. So let's leave it out.

(I see there's a similar test in `ios_project_migration_test.dart`. I suspect that that test's naming as about an "iOS migrator" is a legacy of the `ProjectMigrator` system having originally been called `IOSMigrator`: aa7be00d4be3fffb2cf9fe6f76cde5375ed9abf8 / #71757. It's not really iOS-specific; everything this test would cover is already covered by that one.)

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:35]: Huh — this seems redundant with the initializer this field has in `FakeAndroidProject`. The value that initializer provides will always just get clobbered by this line before anything reads it.

I think the cleanest arrangement is probably to have `FakeAndroidProject` take a constructor argument like `Directory flutterProjectRoot`, and then it can use `flutterProjectRoot.childDirectory('android')` for its `hostAppGradleRoot`. And these lines can say:
```dart
    project = FakeAndroidProject(flutterProjectRoot: memoryFileSystem.root.childDirectory('app'));
```

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:48]: Huh, there are two different loggers involved here. That feels fragile — like it'd be easy for a test to wrongly pass because, for example, an `isEmpty` check was made on one logger, but the output it was meant to detect actually appeared on the other logger.

It looks like in `ios_project_migration_test.dart` there's just one logger, and this global `testLogger` isn't used. Having just one logger sounds like a good idea to me.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:32]: The `MemoryFileSystem.test()` constructor seems like it has the cleaner semantics for a test — less susceptible to flaking.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:62]: nit:

```suggestion
        final DateTime previousLastModified = topLevelGradleBuildFile.lastModifiedSync();
```

Shorter, which makes these lines less fearsomely long and easier to read. :-) Also describes more of the pertinent meaning — I think the expectation a few lines below is easier to make sense of when it's comparing "this file's last-modified time" to "the previous last-modified time", rather than to "this file's last-modified time" which sounds like comparing to itself.

**Review by gnprice — COMMENTED** [review]
Thanks for writing this! Generally looks great. Various small comments below.

**Comment by gnprice** [comment]
Would also be great to have a review from someone who's worked with this migration system. @jmagman, tagging you in as the expert, but please feel free to pass it off to someone else.

File: packages/flutter_tools/lib/src/android/migrations/top_level_gradle_build_file_migration.dart
Comment (bartekpacia) [reply packages/flutter_tools/lib/src/android/migrations/top_level_gradle_build_file_migration.dart:45]: It makes sense, thanks.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart]: Yeah, I agree.

I rewrote `FakeAndroidProject` to be:

```dart
class FakeAndroidProject extends Fake implements AndroidProject {
  FakeAndroidProject(this.hostAppGradleRoot);

  @override
  Directory hostAppGradleRoot;
}
```

because the path was overridden anyway in `setUp()`.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart]: done :)

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart]: Done [above](https://github.com/flutter/flutter/pull/122290#discussion_r1141778817).

Thanks!

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:54]: yeaaah, I wasn't happy with that name as well - thanks!

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart]: Fix'd!

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart]: Yup, good catch as well. I removed all these unnecessary excepts.

**Comment by bartekpacia** [comment]
Thanks a lot for the review @gnprice! I addressed all the things you mentioned.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:27]: nit: make this a named parameter — otherwise it's not evident when reading the call site what the meaning of the parameter is supposed to be

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart]: Hmm, `ios_project_migration_test.dart` does have expectations like these on `statusText`, though, so that makes me think they're likely useful. (The difference is they're using the same logger that was passed to the migrator.)

Looking into those, in addition to checking it's empty in some tests, in other tests it has checks like this:
```dart
        expect(testLogger.statusText, contains('Upgrading project.pbxproj'));
```

That in turn looks like a reference to this in the [ProjectMigrator] logic:
```dart
    if (migrationRequired) {
      logger.printStatus('Upgrading $basename');
```

So effectively it's a way of checking that the file was, or wasn't, modified.

In these tests, there's the `lastModifiedSync` checks for not being modified, and when it is modified the change in contents is clear enough. So I'm happy with leaving them out.

**Review by gnprice — APPROVED** [review]
Thanks! LGTM except one nit. We'll also want that second review as mentioned above at https://github.com/flutter/flutter/pull/122290#issuecomment-1475397420 .

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:27]: Right, done.

**Review by jmagman — COMMENTED** [review]
This migration should be run in all the example/integration test projects so when folks run them locally they don't have changes to their working copy.

Here's an example of a migration that did that: https://github.com/flutter/flutter/pull/108331/files

**Comment by jmagman** [comment]
@reidbaker can you add the right person to review?

**Comment by bartekpacia** [comment]
Uh, `Mac web_tool_tests` are failing and I from looking at the logs, I can't figure out why.

**Review by reidbaker — APPROVED** [review]
The gradle parts look good to me that said I a have not yet written a migration. Pinging
@camsim99 to see if they have any migration based input.

**Comment by camsim99** [comment]
> The gradle parts look good to me that said I a have not yet written a migration. Pinging @camsim99 to see if they have any migration based input.

I don't have any experience writing migrators, so I would defer to you @reidbaker.

File: packages/flutter_tools/lib/src/android/migrations/top_level_gradle_build_file_migration.dart
Comment (jmagman) [inline packages/flutter_tools/lib/src/android/migrations/top_level_gradle_build_file_migration.dart:26]: @camsim99 @reidbaker @gnprice I can review the "migration" part (that LGTM) but I don't know anything about the gradle part of this.
This only changes `android/build.gradle` but not `android/app/build.gradle`, do they both need to be changed?

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:91]: This seems unused?

**Review by jmagman — CHANGES_REQUESTED** [review]
Remove dead code in test.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (gnprice) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:91]: Ah indeed, I think that was for this test in a previous version: https://github.com/flutter/flutter/pull/122290#discussion_r1141463972

File: packages/flutter_tools/lib/src/android/migrations/top_level_gradle_build_file_migration.dart
Comment (gnprice) [reply packages/flutter_tools/lib/src/android/migrations/top_level_gradle_build_file_migration.dart:26]: Yeah, changing just the one is correct. That's the only one where the templates happen to contain the pattern that generates the Gradle warning in #122289.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart
Comment (bartekpacia) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart]: Oops, forgot about it.

Removed. Thanks.

**Review by reidbaker — APPROVED** [review]

**Comment by bartekpacia** [comment]
Rebased with `master` in hope it fixes (seemingly) random test failures.

**Review by jmagman — APPROVED** [review]
LGTM

**Pull request merged**
